### ISSUE TITLE
[SUGGESTION] Animated block model hook for TESR's

### ISSUE THREAD

**Issue by AlexIIL** [MinecraftForge/MinecraftForge#2163]
This would allow simple blocks to be animated by extension of the current block model system. Essentially, this would change block model loading to have both the static chunk renderer (for all the non-changing elements, like now), and a TESR hook for all the changing (animating, moving etc) elements.

All of the examples are partial implementations of what an engine renderer looks like (copied mostly from buildcraft, but ignoring the different types and stages).

The first changed part would be the block state loading, which would have an additional "variables" object (containing a list of the variables used and their default values). In each of the variants an additional "animated" tag would determine whether it should render via a TESR (and get the values for the variables on each frame) or in normal block rendering (getting the values from the defaults given above).

````json
{
  "forge_marker": 1,
  "variables": {
    "progress": 0
  },
  "variants": {
    "facing":{
      "up":    { },
      "down":  { "x":180 },
      "east":  { "x":90, "y":90 },
      "west":  { "x":90, "y":270 },
      "north": { "x":90 },
      "south": { "x":90, "y":180 }
    },
    "moving":{
      "false": { "model":"example:engine_model" },
      "true":  { "model":"example:engine_model", "animated":true }
    }
  }
}
````

The second change would be to the model loading system, which would have to be changed to allow for expressions in place of both the positions and UV coordinates.

````json
{
	"elements": [
    {   
      "from": [ 0, 0, 0 ],
      "to": [ 16, 4, 16 ],
      "faces": {
        "down":  { "uv": [ 0, 0, 16, 16 ], "texture":"#base" },
        "up":    { "uv": [ 0, 0, 16, 16 ], "texture":"#base" },
        "north": { "uv": [ 0, 0, 16, 4 ], "texture":"#front" },
        "south": { "uv": [ 0, 0, 16, 4 ], "texture":"#front" },
        "west":  { "uv": [ 0, 0, 16, 4 ], "texture":"#front" },
        "east":  { "uv": [ 0, 0, 16, 4 ], "texture":"#front" }
      }
    },
    { 
      "from": [ 0, "progress + 4", 0 ],
      "to": [ 16, "progress + 8", 16 ],
      "faces": {
        "down":  { "uv": [ 0, 0, 16, 16 ], "texture":"#base" },
        "up":    { "uv": [ 0, 0, 16, 16 ], "texture":"#base" },
        "north": { "uv": [ 0, 4, 16, 8 ], "texture":"#front" },
        "south": { "uv": [ 0, 4, 16, 8 ], "texture":"#front" },
        "west":  { "uv": [ 0, 4, 16, 8 ], "texture":"#front" },
        "east":  { "uv": [ 0, 4, 16, 8 ], "texture":"#front" }
      }
    },
    {
      "from": [ 4, 4, 4 ],
      "to": [8, "progress + 4", 8 ],
      "faces": {
        "north": { "uv": [ 4, 0, 12, "progress" ], "texture":"#chamber" },
        "south": { "uv": [ 4, 0, 12, "progress" ], "texture":"#chamber" },
        "west":  { "uv": [ 4, 0, 12, "progress" ], "texture":"#chamber" },
        "east":  { "uv": [ 4, 0, 12, "progress" ], "texture":"#chamber" }
      }
    },
    {
      "from": [ 4, 4, 4 ],
      "to": [  12, 16, 12 ],
      "faces": {
        "down":  { "uv": [ 0, 0, 8, 8 ], "texture":"#trunk" },
        "up":    { "uv": [ 0, 0, 8, 8 ], "texture":"#trunk" },
        "north": { "uv": [ 8, 4, 16, 12 ], "texture":"#trunk" },
        "south": { "uv": [ 8, 4, 16, 12 ], "texture":"#trunk" },
        "west":  { "uv": [ 8, 4, 16, 12 ], "texture":"#trunk" },
        "east":  { "uv": [ 8, 4, 16, 12 ], "texture":"#trunk" }
      }
    }
  ]
}
````

The final addition would be a hooking class to allow any TESR object to get the animated section, change the variables, and render off the animated model.

````java
package example;

// Imports missing for the sake of an example.

public class EngineRenderTESR extends TileEntitySpecialRenderer {
  private final AnimatedBlockRenderer engineRenderer;
  private final Variable progressVariable;

  public EngineRenderTESR() {
    // An example class would get the model by its blockstate resource location
    engineRenderer = BlockModelHandler.getAnimatedModelRenderer("example:blockstates/engine_blockstate");
    /* Note that "progress" is the same as in engine_model.json.
      In this example usage, variable are obtained by their names first, and then updated each time the renderer is called.
      I'm not sure about actual speeds, but this would probably be faster than looking up a string every render tick,
      and safer than using an underlying double array. */
    progressVariable = engineModel.getVariable("progress");
    /* The above calls would probably throw an exception if either of them did not have the correct arguments (either the
      block model file did not exist, or the varaible did not exist) as it would probably be a coding error to pass 
      incorrect arguments. However, it is possible that a resource pack creator doesn't use some of the variables in their 
      model file, so it would might be better to just print an error message, and return null. */
  }

  @Override
  public void renderTileEntityAt(TileEntity tile, double x, double y, double z, float partialTicks, int wtfIsThis) {
    TileEngine engine = (TileEngine) tile;
    // This would obviously be handled differently depending on the mod
    double progress = engine.getProgress(partialTicks);
    IBlockState state = tile.getWorld().getBlockState(tile.getPos());
    /* Because the example implementation of "moving" block state is likely to be outside of the metadata, we probably 
      need to get it */
    state = state.getBlock().getActualState(state, tile.getWorld(), tile.getPos());
    /* For simplicity, all variables are double's. I currently can't think of a situation where they would need to be 
      intergers, however an integer cast function could be created. It is possible that they would need to be strings 
      (for textures, for example rendering different fluids in a tank) but they would be handled differently as 
      you can't really do anything except concatenate strings. */
    progressVariable.set(progress);
    /* Manual translation with OpenGL. This translation could be done by the forge renderer itself, but as this suggestion is 
      about making things accessable to resource pack creators and pulling block models together with the animated parts, 
      I am not suggesting to remove direct OpenGL usage. A seperate method called renderOffset() could take in the state and
      the positional offset to make things simpler if all the TESR needs to to is render off the animated block model. */
    GL11.glTranslated(x, y, z);
    // The renderer needs the block state to determine what should be animated 
    // (Note the "animated":true property in the blockstate file)
    engineRenderer.render(state);
    GL11.glTranslated(-x, -y, -z);
  }
}
````

I am going to make a pull request for this, however I would like to check to see what the current direction forge is going in terms of block rendering, and whether I should wait for forge for 1.8.8 to start implementing it.

A seperate (but related) suggestion is a way to get an entire model from a location and render it off inside both a TESR and a custom block model (so, provide a way to get the list of the block models, and apply arbitary rotation to all of them). However I have not tried (very hard) to find this, so you can probably ignore this.

Also if this is NOT wanted in forge it would be great to know quickly, so I can implement this now for BuildCraft, rather than waiting for a "no".

**Comment by asiekierka** [comment]
+1

**Comment by RainWarrior** [comment]
This is a great idea, and something along these lines will surely be available at some point in the future.
However, there are a couple of points I'd like to make:

1. Vanilla JSON models have some strict limitations imposed on them - limited rotations/scaling/e.t.c; and, extending them is tricky due to the need to burden of maintain all the patches/rewriting everything, and keeping the rewrite in sync with the vanilla; therefore, I'd go for the other approach - implementing 3rd party model formats, without those limitations, and with animation support built-in. This also gives the benefit of the person doing the models using the model editor of their choice, provided that it can export said model format.

2. "variables" section: keeping in mind the first point - there's a notion of animation tracks and keyframes in the modelling world, and this concept is very close to that.

3. Forge has the static infrastructure of this already in - IModel is the model with the "animation variables" that can be changed, and IModelState can represent those animation variables. There's no implementation of that in the JSON yet, since there's no clear overview of what different model formats need to support, and whether there's anything more specific that they have in common. It'll be nice to figure it out - @AtomicBlom is working on porting his OGEX loader to 1.8, I've implemented [B3D](https://github.com/MinecraftForge/MinecraftForge/tree/master/src/main/java/net/minecraftforge/client/model/b3d) loader that's in Forge, and you need something that will suit BuildCraft.

The 1.8.8 is where the focus of the Forge developement is moving to, I think, after the recent rendering changes are stabilized, but the model stuff shouldn't change very much for that - main difference will be reintroduced generics.

As for the getting the model form the location - you can get it by using BlockRendererDispatcher.getModelFromBlockState.

TL;DR: yes, this is definitely needed, but more discussion and collaboration would be great :Р

**Comment by AlexIIL** [comment]
One of the main reasons for this was to expose more of the rendering code to resource pack authors, so using a different format makes a lot of sense. Ideally this would be left up to the user (not the modder) to use whatever model format they wanted, so in the blockstate or model JSON they could specify which file to take from (rather than require the modder to explicitly select the model type to use).

The variables exists to allow essentially extended block state variables to be taken from the tile itself, which may be the similar to as keyframes.

I would still quite like to use JSON in some way (Although I haven't used a model program before, so my opinion might change on that).

**Comment by AlexIIL** [comment]
Looking around minecraft+forge I don't actually think what I was after makes a lot of sense- get a json model (and JUST a model) from a resourcelocation (independant of a block) and use that as a sub-part. Basically getting a sub-model.

**Comment by rubensworks** [comment]
@AlexIIL I think your suggestion for loading models from just json instead of a blockstate makes a lot of sense. I would benefit from such an addition as well, since I currently require making fake-blocks everytime I need to load such a model. Maybe a separate issue should be made for this?

**Comment by RainWarrior** [comment]
@rubensworks is ModelLoaderRegistry.getModel what you're looking for?

**Comment by rubensworks** [comment]
@RainWarrior You can ignore my comment, `ModelLoaderRegistry.getModel` does the trick indeed.
I actually required direct loading of *blockstate* json files (it's been a while since I worked on this code...).
But I'll make an issue soon to explain my problem, as I want to reuse the capabilities of blockstates, without needing to make a new block for this (if that makes any sense).

**Comment by AlexIIL** [comment]
@rubensworks Did you ever open an issue about using a blockstate without a block? Its something BuildCraft will definitely want so I'm interested in if you have made any progress with it.

**Comment by AlexIIL** [comment]
In terms of progress with my current work, its at a stage where I have successfully implemented most of the animated model elements required, but I still need to tie models together with the actual blockstate (hence my previous comment) and with the ModelLoaderRegistry -but I have outlined my thoughts at [linky link](https://github.com/AlexIIL/MinecraftForge/blob/89f35508ace3ebcc6c4292b0ff63b649543f995d/src/main/java/net/minecraftforge/client/model/anim/AnimatedModelRegistry.java#L20-L37).

I have deviated from my original plans by instead creating a separate model format (json)- but similar to avoid patch issues. One thing that hasn't happened is an "animated" tag inside of the blockstate.

If there hasn't been any plans for loading arbitrary state information (as in, without a block but otherwise parsed in a similar manor to the current forge block state), and such a thing would want to be in forge, then it would be nice to create a separate implantation.

From looking at the latest 1.8.8 dev, it feels like I am not heading in quite the same direction as that (not being able to clone the 1.8.8 repo to test doesn't help, but that’s a separate problem) specifically towards using animated models. Currently all animated models have the parent interface ```` IAnimatedModel```` that extends ````IModel```` and ````IModelElement```` (The latter being simalir to ````IModelPart```` but not as it has a few methods that control rendering and texture dependencies). All animated models are based around ````AnimatedTransformation````- a (kindof) reimplementation of ````TRSRTransformation```` that allows for any number of seperate transformations (which MAY help performance as all of these are evaluable functions that need to be evaluated any time a variable changes).

One thing I did notice (and why I don't just evaluate to a ````TRSRTransformation````) is that I could never get rotations working in the way I expected them to. This honestly feels like I am missing how they are interpreted and used in a matrix.

**Comment by RainWarrior** [comment]
Yes, this is not quite the direction I want to go, but I think what I have in mind should yield similar results.
I'll try to outline the idea in the next couple of days, it's still not quite formalized.

**Comment by rubensworks** [comment]
@AlexIIL Thanks for reminding me. Created it as #2198.

**Comment by AlexIIL** [comment]
@RainWarrior Out of interest, is your plan for animations formalized now, or do you want more time? or should I try to change my system to play well with whats currently in forge 1.8.8?

**Comment by RainWarrior** [comment]
@AlexIIL I want a bit more time :P

**Comment by RainWarrior** [comment]
To clarify: I want to have something that would:

1. be usable for everything - blocks, items and entities
2. have a high-ish level API - not just variables, but also animation states and transitions
3. also be reasonably fast
4. would be applicable to all custom models too, without them being aware of the variable system

I have some thoughts as so what that would look like: here's the preliminary gist with what I would like to see from the corresponding json files: https://gist.github.com/RainWarrior/5cd13821c7b53ae771af
(that gist will probably change quite a bit before I think it's good enough)

Right now I'm trying to see what the API for this would look like in code, and if it's possible to make it fast enough with the structure like this.

**Comment by AlexIIL** [comment]
That's a lot better than what I originally thought of :)

I was playing around with performance with my own implementation, and what took the most time was just rendering everything out to opengl (I tested it with a more complex variant of an engine, with 1331 arranged in an 11x11x11 cuboid). This was transforming each point with a matrix every frame for every model (for every TESR). I would assume that it would be possible to make it fast enough (Caching static models into a vertex array and multiplying the GL matrix by my own made more difference than anything I tried to do to speed up calculation).

**Comment by laci200270** [comment]
@RainWarrior maybe you can add texture translation option?(chaning the opacity of textures lineary and cut texture with a percentage)?(FOr example energy cells status indicators)

**Comment by AlexIIL** [comment]
@laci200270 That's something that BuildCraft already needs, so if I am involved it will be added- specifically to do with texture and positional changes. Texture opacity is a different matter, and probably not too hard to add though.

**Comment by AlexIIL** [comment]
Adding to @laci200270 suggestion, it would probably be helpful to change all 4 colour elements (RGBA) of the texture as well (Specifically for custom rendering fluids, as they have a colour associated with them).

**Comment by RainWarrior** [comment]
animation/engine.json in the gist is the part I'm least worried about - animation stuff wouldn't care about the specifics of the concrete model format (vanilla JSON models, in this case), and the model would be able to change anything inside the animation clip - position, colors, e.t.c; animation tracks for vanilla models is a clean separate part, much like the vanilla model loader in the current model system.

**Comment by laci200270** [comment]
Any ideas when this will be done?

**Comment by RainWarrior** [comment]
Should be implemented by #2331.
